### DISCUSSION TITLE
How am i able to commit and push a 5GB File in Windows?

### DISCUSSION THREAD

**Discussion by dhiwakarK** [git-lfs/git-lfs#4567] [Q&A] [answered]
How am I able to commit and push a 5GB File in Windows? I thought the 4 GB limit continues to persist in Windows due to the known issue. The [FAQ](https://github.com/git-for-windows/git/wiki/FAQ) continues to speak about this issue, and there even exists a pending [pull request ](https://github.com/git-for-windows/git/pull/2179)on this issue 

**Repro Steps**
1. Created a fresh *Azure Windows 10 Pro VM*
2. *Configured the VM with the following* all latest as of date i.e. 3rd of August 2021
- [Git for Windows](https://git-scm.com/download/win)
- [Visual Studio Code Insiders](https://code.visualstudio.com/insiders/)
- [Git LFS](https://git-lfs.github.com/)
- [Python](https://www.python.org/downloads/)
- Added some Jupyter Python Extensions for Visual Studio Code
3. _Generated a dummy file_ with 5 GB data
![image](https://user-images.githubusercontent.com/18525045/128173369-d8af5957-7a8e-4bf8-bb18-a2e6053a100c.png)
4. _Zipped the above file_ - so that I can check file corruption. Size continues to remain above 4 GB
5. _Created a new repository in a Test GitHub Enterprise Instance_, cloned the same and executed the following commands in order
6. `Git LFS Install`
7. `GIT LFS Track "*.zip"`
8. Confirmed that the .gitattributes files are generated
9. Added the 5GB zip file
10. Confirmed through Git status that the file exists
11. `Git Add`
12. `Git Commit`
13. `Git Push`
![image](https://user-images.githubusercontent.com/18525045/128173680-f2935541-48f4-41f3-8e26-b1363e49f1d9.png)
14. _Note that the Git LFS is in action_
![image](https://user-images.githubusercontent.com/18525045/128173814-6b2f79e2-52a3-4e21-a566-9276fcad9e4b.png)
15. To proceed further I even did a checkout of a fresh branch, pushed some changes
![image](https://user-images.githubusercontent.com/18525045/128174267-ee9b0f9b-1148-4660-bc82-609c6e87705e.png)
16. Zip file is NOT corrupted, and there were NO errors

**So, big question, what am I missing? Why was I allowed to work on a 5GB file ** 😃 

**Does it mean the 4GB file limit doesn't apply all the time? Is there a way I can reproduce the issue with certainty?**

👀 @PhilipOakley @dscho

**Comment by dscho** [comment, marked as answer]
This behaves as intended. Git for Windows' 4GB limit is for contents tracked inside Git's own database. With Git LFS, only a short and sweet hash is tracked in that database, and the actual file contents are tracked by _Git LFS_.

Therefore you can use Git LFS as a workaround if you need to track larger files with Git for Windows.

**Reply by dhiwakarK** [reply]
Then why is the [GIT-LFS FAQ ](https://github.com/git-lfs/git-lfs/wiki/FAQ)saying otherwise (see screenshot below)? As you see in my screenshots above I'm not even using any smudge configurations or Git LFS Push..but it still works.

![image](https://user-images.githubusercontent.com/18525045/128201975-3a155017-4035-4d38-8a39-f7a5323ccdec.png)

**Reply by bk2204** [reply]
It may be that running `git add` happens to work, but it doesn't when you smudge files into the working tree.  To reproduce, you'd need to check out a branch that doesn't have the large file at all and then switch back to one that does.  When you do that, your file will be truncated by Git.

**Reply by dhiwakarK** [reply]
Got it 👍🏼  

![image](https://user-images.githubusercontent.com/18525045/128305200-5684b034-cf6c-4c83-aa3e-1a63904b9f31.png)

File was corrupted a well..
![image](https://user-images.githubusercontent.com/18525045/128305310-d60ed041-86cd-45a6-98ef-2528f3088991.png)

As documented, after setting the smudge filters it works fine. Even the LFS Pull works fine, and the files are not corrupted anymore.

![image](https://user-images.githubusercontent.com/18525045/128306788-4f62cb31-1500-4ba3-9e62-d91fde455cd1.png)

![image](https://user-images.githubusercontent.com/18525045/128306863-58198fe7-8844-44ff-b44f-5aa8f7cdc648.png)

**Reply by dscho** [reply]
Oh, sorry about my incorrect answer, then!

**Reply by dhiwakarK** [reply]
All good mate, thanks for all the prompt answers. Really appreciate it!

**Reply by xls** [reply]
We have confirmed that using git lfs directly works all good, and with smudge and filter skip all is good.

Although a ok work-around for now, third-party tools like embedded git clients (visual code, visual studio, gitkraken etc). will not perform a git lfs pull automatically. A normal git pull will fetch the "reference files" so folder look all good but files are just xml references to the lfs repo.

I can see the mass confusion among our dev / test / compliance when they check out a game, and files are there, but not the actual files. and that git lfs pull is required.


The issue is related to c/c++ long type is 32bit on Windows (independent of CPU architecture) -  while on 64bit linux it is 64bit.

A fix was proposed 2 years ago (https://github.com/PhilipOakley/git.git branch: size_t6).

The problem is in git client, both in crc32 (checksum) and in many of the file io functions (see branch).

it is a bit crazy that this has not been prioritized by community.

**Reply by dscho** [reply]
> A fix was proposed 2 years ago (https://github.com/PhilipOakley/git.git branch: size_t6).

This proposed fix will need a ton of work to get over the finish line. Enough work to keep several contributors busy for quite a while.

> The problem is in git client, both in crc32 (checksum) and in many of the file io functions (see branch).
> 
> it is a bit crazy that this has not been prioritized by community.

I have to admit that I am somewhat allergic to statements like these. Whenever anybody talks about "the community" as if they were not part of it, and pretend that the responsibility for a certain fix they desire so much lies not with them, but with others, all too often words like "crazy" are used with the obvious goal of shaming _others_ into working on the fix.

This elicits unnecessary negative emotions, naturally provoking backlash (such as questioning the choice to put such large files into a source code repository).

An obvious consequence is the engineers like myself don't want to come near these conversations.

**Reply by xls** [reply]
@dscho my apologies, by community I mean the git users operating under windows and.. by crazy I mean I am surprised that there is no more interest to support LFS 4gb+ files (respectfully it does work just fine under linux). My background is with game studios were large assets are frequently witnessed as a dependency to projects.. Of course these files would typically live in repositories designed to handle larger files such as LFS; while game studios are technically savvy and can work around the problem with usually complex build and asset baking systems. The easy of use with git+LFS  should be of interests for this segment.
again, no disrespect to hard working community.

**Comment by PhilipOakley** [comment]
@xLs @dhiwakarK Are you, your colleagues, or companies, in a position to provide some effort as part of the community? (the actual community for Windows issues is rather small)

As said, it will need '[someone(s)](https://www.lollydaskal.com/leadership/story-everybody-somebody-anybody-nobody/)' to assist in looking at the various trip hazards along the road to implementation. 

The basic end points have been identified (Zlib, config, FS_funcs), but finding a clear and obvious pathway (small steps) to implementation needs a lot more insights into getting past the ubiquitous coercion issues between the differing pointer length systems, and the size (loc) of the code base. 

IIUC, the same issue affects the Raspberry Pi, which has the same [LLP32](http://nickdesaulniers.github.io/blog/2016/05/30/data-models-and-word-size/) system.

The single mega-patch method will not be acceptable.

As it is such a big change its now obvious that it's not something I'll try on my own.

**Reply by dhiwakarK** [reply]
👋🏼 @PhilipOakley Can we discuss this in the Git forum, instead of the LFS Forum? 

Here are **my observations** and I'll let @PhilipOakley @dscho pitch in to provide additional information/clarity. From my understanding, it looks like the [changes proposed](https://github.com/git-for-windows/git/pull/2179) are very large. @dscho suggested like below
>The trick is to break this PR down into bite-sized chunks so that we can benefit maximally from code review (if the patches are too big, you will not benefit from that extra safety).

@PhilipOakley, are you in a position to split this change into multiple or would you need help there itself - even a draft will help warm up the discussion? I'm no expert, but looking at the Pull Request discussion I would start the discussion around how this major change should be addressed (in terms of chunks) rather than proposing the code change itself. 

From the community involvement, there are a couple of points ...
- On one side, I could see demand picking up. Example: Machine learning models & sample datasets, Gaming Development Studios, etc.
- People have started their own implementations to handle large files Example DVC.org
- I'm not sure how large this need is considering the general recommendation to not store large files in the repository, and considering the fact that the other solutions exist to fill the gap. 
- BUT, the community is the best judge to see how big this need is and hence I proposed you to start the discussion around how this change should be addressed. 

Just my thoughts 🙇🏼

**Comment by PhilipOakley** [comment]
Hi @dhiwakarK, Thanks for the reply. 

In part, my reply here, was because the "Large File System"  may well benefit from resolving the Windows >4Gb limit (e.g. #2434). So there may be some users here who could support the Windows (or mainline Git) update to ensure 64 bit cross platform usage (e.g. RaspPi & Windows)

Happy to continue discussions over at  https://github.com/git-for-windows/git/pull/2179

There has been on and off discussions over the years, e.g. https://github.com/git-for-windows/git/issues/1063 (closed), 
https://github.com/git-for-windows/git/issues/1848 (open?)
https://github.com/git-for-windows/git/pull/2166 (closed)
https://github.com/git-for-windows/git/pull/2179 (open)
along with https://github.com/git-for-windows/git/pull/3306 which looks at making sure that it's real easy to open GfW in Visual Studio to do investigations using such Windows tools.

At the moment I don't see an actionable way of splitting up the very large middle part of the needed changes. Dscho hasn't linked to anything, other than saying it needs doing, and he had a go at it. The zlib/crc32 fizes are sorted, along with the extra config_type and File System routines, but the ripple effect of the `long` vs `size_t` difference on Windows, through out the codebase, is massive (C4267 below). See [Why size_t matters](https://www.embedded.com/why-size_t-matters/) for the LLP codes. 

```
1912 warnings; 0 errors
C4005 (2) UNICIDE macro redefinition
C4018 (4) signed/unsigned mismatch
C4133 (4) Function incompatible types
C4267 (1891) /size/ conversion, possible loss of data
C4334 (4) '<<' result if 32bit shift implicitly converted to 64bit - intended?
C4477 (1) fprintf format %s requires char but WCHAR given (variadic arg 1)
C4996 (5) 'GetVersion' deprecated, gethostbyname use getaddrinfo[W] ..
unkwn (1) CMake : translations won't be built..(expected)
```

So, discussions at https://github.com/git-for-windows/git/pull/2179

**Reply by dscho** [reply]
> At the moment I don't see an actionable way of splitting up the very large middle part of the needed changes.

The trick is to try not so much splitting up the very large middle part, but to _redo_ them, incrementally. That is, you need to pick _one_ data type that you want to change from `unsigned long` to `size_t`. Just one. It should be as peripheral in the code as possible, so that you don't have to change a lot of code in order to make things compile. In other words, I would still suggest precisely what I suggested almost 2 years ago: https://github.com/git-for-windows/git/pull/2179#issuecomment-527463561

**Reply by PhilipOakley** [reply]
> The trick is to try not so much splitting up the very large middle part, but to _redo_ them, incrementally. That is, you need to pick _one_ data type that you want to change from `unsigned long` to `size_t`. Just one. It should be as peripheral in the code as possible, so that you don't have to change a lot of code in order to make things compile. In other words, I would still suggest precisely what I suggested almost 2 years ago

Unfortunately this fails to communicate what I suspect you think it does. It still sounds like hand waving with a lot of presumptions.

Do you still have your example?

**Reply by dscho** [reply]
> Do you still have your example?

You mean the proof-of-concept commits? Yes, they are still where I pushed them: https://github.com/dscho/git/commits/size_t6

**Reply by dscho** [reply]
Note that https://github.com/dscho/git/commit/9c2f8de134713f7e04fc2402567c5b0c29605737 still needs to be re-done in the indicated way.

**Reply by PhilipOakley** [reply]
> the proof-of-concept commits? Yes, they are still where I pushed them: https://github.com/dscho/git/commits/size_t6

Thanks, I hadn't realised that you had updated the branch contents. I was initially confused at my sample authorship.

I still have concerns about how to move from singleton changes to a fully working set of changes, e.g. tests & checks for completeness, but that can be discussed at  git-for-windows/git#2179.

**Comment by rcoup** [comment]
FWIW fixing the described LFS 4GB Windows file limitation has now been [submitted to core Git](https://github.com/gitgitgadget/git/pull/1068), and @dscho is intending on [including it in Git for Windows v2.34.0](https://github.com/git-for-windows/git/pull/3487).

**Reply by dscho** [reply]
@rcoup thank you for looping back here!

**Reply by rcoup** [reply]
Happenstance! A bit of time spelunking through various LFS issues and PRs this morning to end up discovering it was submitted today! 😃 🤞

**Reply by xls** [reply]
well done @dscho, @vtbassmatt, @PhilipOakley  and everyone involved.. thanks again!
